### ISSUE TITLE
RuntimeEncoder and RuntimeDecoder produce invalid SymbolicPulses with terra 0.22

### ISSUE THREAD

**Issue by wshanks** [Qiskit/qiskit-ibm-runtime#657]
**Describe the bug**

Using the RuntimeEncoder and RuntimeDecoder on a circuit with a gate calibration produces invalid pulses when the calibration includes one of the SymbolicPulse shapes included in qiskit-terra.

**Steps to reproduce**

```
import json

from qiskit import QuantumCircuit, assemble, pulse
from qiskit_ibm_runtime.utils import RuntimeEncoder, RuntimeDecoder


with pulse.builder.build(name="x") as sched:
    pulse.play(pulse.Gaussian(100, 0.8, 10), pulse.DriveChannel(0))

    
qc = QuantumCircuit(1)
qc.x(0)
qc.add_calibration("x", (0,), sched)

dumped = json.dumps(qc, cls=RuntimeEncoder)
loaded = json.loads(dumped, cls=RuntimeDecoder)

print(type(qc.calibrations["x"][((0,),())].instructions[0][-1].pulse))
print(type(loaded.calibrations["x"][((0,),())].instructions[0][-1].pulse))
```

This prints:

```
<class 'qiskit.pulse.library.symbolic_pulses.SymbolicPulse'>
<class 'qiskit.pulse.library.symbolic_pulses.Gaussian'>
```

**Expected behavior**

Both versions of the pulse should be `SymbolicPulse`.

**Suggested solutions**

I believe that #577 would fix the problem though I did not test it. I did test that doing a direct dump and load with qpy from qiskit-terra produced the correct class for the pulse.

The issue is that qiskit-terra 0.22 deprecated the SymbolicPulse subclasses by changing them to define a `__new__` constructor that generates a `SymbolicPulse` object directly. The 0.21 qpy loader creates the SymbolicPulse subtype directly with `object.__new__()` and sets its attributes without invoking the class's constructor. With the changes in 0.22, this loader now generates a metaclass that does not function as a proper SymbolicPulse class.

**Additional notes**

@kevinsung ran into this issue trying to run a job with pulse gates using qiskit-terra 0.22. What I think happens is that when the backend code loads the qpy payload and assembles it into a qobj, it hits [this conditional](https://github.com/Qiskit/qiskit-terra/blob/dd7f9390cf076f9545fa6086de486f2c06764ea5/qiskit/assembler/assemble_schedules.py#L228) that checks for a `SymbolicPulse` and skips over it, skipping pulling the pulse parameters out of the object. The backend ends up returning a "Pulse name is not a string" error because the pulse name is left as `None`.

One thing I think might merit further discussion is how to avoid this kind of issue. I believe that #577 is backwards compatible -- I *think* it will work even with qiskit-terra 0.21 (and maybe earlier). However, the changes made in qiskit-terra 0.22 were not compatible with the 0.21 qpy loader in the runtime package. Is there a way to update qpy in terra without breaking the runtime package using the previous version? Would it help if qpy were its own package and terra committed to being compatible with more than one version at a time? @jakelishman @mtreinish @nkanazawa1989 

Final note -- there is also #655 which is a similar symptom of the same issue, I think, but runs into a different error. 

**Additional Information**

- **qiskit-ibm-runtime version**: 0.8.0
- **qiskit-terra version**: 0.22.3
- **Python version**: 3.10.6
- **Operating system**: Fedora 36

**Comment by kevinsung** [comment]
Possible workaround suggested by Will:
> The best you might be able to do is convert all the pulses into Waveforms. Another option is to go through and overwrite _pulse_type on the pulses so that it is not Drag or GaussianSquare. Then the pulses will get serialized as parametric pulses (instead of waveform) but in the runtime when they are deserialized they will not be recongized pulse types and will be converted to Waveforms.

**Comment by jakelishman** [comment]
Terra already commits to complete backwards compatibility with QPY, so new Terra will always told QPY generated by old Terra. What you're suggesting here is _forwards_ compatibility, i.e. that we should constrain Terra's development so much that old Terra can load new objects. That's not reasonably possible for a library that needs to evolve like Terra.

The solution to me is just for Runtime to update their vendored version of QPY a little faster. Given it's vendored for stability, and Terra's backwards compatible guarantees, Runtime could proactively update QPY from Terra main, but at the moment it's still using QPY loading code from Terra 0.21, while telling people to use 0.22 to generate the circuits.

**Comment by wshanks** [comment]
> What you're suggesting here is forwards compatibility, i.e. that we should constrain Terra's development so much that old Terra can load new objects that might not have existed when the loader code was written (like this case).

No, that is not quite what I was suggesting. In particular, you reference "old Terra" but the issue we are talking about here involves new terra on both the client and the server (I used terra 0.22 and runtim 0.8.0 and I believe the server is using those as well). "qpy" is both a binary format for storing quantum circuits and a Python library for storing and loading the binary format. One way to describe the problem is that qpy the Python library is developed as part of terra and is tested for compatibility within terra and for compatibility with loading previous versions of the binary format, but these tests are always of the qpy Python library integrated into a version of terra. With qiskit-ibm-runtime and qiskit-ibm-provider, we have the qpy Python library effectively split out as a separate package from terra, and that opens up the question of compatibility between the qpy library version 0.22 and terra version 0.23 for example. (My prevous suggestion was to test terra 0.23 with the 0.22 qpy library somehow).

> The solution to me is just for Runtime to update their vendored version of QPY a little faster.

I think this is almost right. What I think could mostly solve the issue is if on the server the runtime used terra's qpy library to load the payload rather than the vendored runtime one. I think the purpose of the vendored qpy is to prevent a user updating terra before the runtime service's version of terra is updated and generating a qpy payload with a qpy version number higher than the service can load. So the runtime package holds back the version of qpy until the version of terra in the runtime service has been updated (that's what I interpret from https://github.com/Qiskit/qiskit-ibm-runtime/pull/577#issuecomment-1279252515). Since new versions of the qpy library should be able to load old versions of the binary format, I don't think the version of the server needs to be held back.

Using the terra qpy on the server still leaves the possibility that the client could update terra and find an incompatibility between the new terra and the old qpy library for serializing a payload, but at least in this case the user might just accept that even though a new terra has been released they need to wait for the server to support the new feature they tried to use. I could imagine a system where terra's qpy supports generating payloads in older qpy versions and then the runtime queries which version of qpy the server supports and generates a payload of that version with terra rather than using a vendored qpy. That might help avoid issues with incompatibilities due to code refactoring between the vendored qpy and the new terra but could not handle the case where you want to use a feature that only received support in the new version of qpy (and would probably be a pain to maintain -- distinguishing between code changes that can be adapted back to the previous version of qpy and those that are new and require the new format could be difficult).

**Comment by jyu00** [comment]
Runtime started out using terra qpy for both server and client sides, but we had a number of cases where the server could not be updated in time (usually due to deployment issues). This caused jobs serialized with new qpy to fail. As a result we switched to using vendored qpy.

Like @jakelishman mentioned, runtime really should update its qpy more quickly. In this particular case, the update was delayed to avoid potential disruption to Summit work (and kind of fell through the cracks after).

@wshanks made a good point that runtime server doesn't need to use vendored qpy, since qpy is backward compatible and terra qpy will always be ahead of vendored qpy. This would make the server upgrade easier and thus speed up the process.

> I could imagine a system where terra's qpy supports generating payloads in older qpy versions ... rather than using a vendored qpy

This would make it easier for runtime, but I'd image it'd a bit of work for terra.

**Comment by ehchen** [comment]
bump

**Comment by kt474** [comment]
this will be fixed by #577

**Comment by jakelishman** [comment]
Since I'm here again (I'd completely forgotten about this): I think I get what you're saying a bit better now Will.  If I've got it right, what you're suggesting doesn't come to _arbitrary_ forwards dependence, but it still ends up in us trying to get old versions of QPY to be able to load _some_ new objects or objects whose properties have changed slightly, which we usually can't do without updating QPY code.

If that's the case, then I don't think a separate package would be much different from the current situation. We'd still be dependent on Runtime upgrading a dependency that they'd presumably have pinned, which is very similar to updating a vendored version.

**Comment by kt474** [comment]
fixed by #577
